### ISSUE TITLE
keyboard input is being intercepted/blocked by emscripten/sdl

### ISSUE THREAD

**Issue by Elijah23Johnson** [chase-moskal/webp-hero#18]
Disables Input Field's For Safari:

Im running a SSR React app with client side routing. I run
`        const webpMachine = new WebpMachine();
        webpMachine.polyfillDocument()`
everytime a route changes. Either way it freezes input elements and prevents user input. Is there any known work around for this?

**Comment by chase-moskal** [comment]
i think the problem is that the decoding happens on the main javascript thread, and is blocking the ui

the solution would be to workerize `webp.js` so that it runs in a background thread

strategy to implement workerization:

- let's look at some code from the es-module: [`webp-hero/libwebp/dist/webp.js` line 22](https://github.com/chase-moskal/webp-hero/blob/master/libwebp/dist/webp.js#L22)
    ```js
	    this.Module = Module
	    this.webpToSdl = Module.cwrap("WebpToSDL", "number", ["array", "number"])
	    this.setCanvas = function(canvas) { Module.canvas = canvas }
    ```
- we want to workerize the above module, so it operates using worker postMessage api and such
  - it would utilize [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) to perform decoding
  - the module would expose some new worker-available functionality (probably a `decode` function) which internally calls `this.setCanvas(new OffscreenCanvas(x, y))` prior to calling `this.webpToSdl()`
  - then we'd change [`webp-hero/source/webp-machine.ts`](https://github.com/chase-moskal/webp-hero/blob/master/source/webp-machine.ts#L40) to utilize the workerized module via the postMessage api
  - unfortunately, browser support for `OffscreenCanvas` is poor, however we might be lucky if a polyfill like [offscreen-canvas](https://github.com/ai/offscreen-canvas) can do the job

some additional notes about the built `webp.js` module, for anybody looking to do the work:

- `webp-hero/libwebp/` has a dockerfile that builds google's libwebp source into javascript modules (in `webp-hero/libwebp/dist/`)
- in the docker container we have a shell script wrap google's code in es-module and also common-js module formats
  - [`webp-hero/libwebp/source/container-script.sh`](https://github.com/chase-moskal/webp-hero/blob/master/libwebp/source/container-script.sh#L9)
- workerizing webp-hero would involve some work in the [`webp-hero/libwebp/source/`](https://github.com/chase-moskal/webp-hero/tree/master/libwebp/source) area, changing files like `wrapping.start.js.txt` and `wrapping.end.js.txt`
- if the workerized version doesn't work for some of the browsers we care about (like ie11), then we should make this a progressive enhancement by generating a new `webp.worker.js` -- otherwise if the workerized version seems to work for everyone, it shall replace `webp.js`

pull requests are very welcome, as i cannot get to this soon

**Comment by chase-moskal** [comment]
@PosSoolutions 

> I run `const webpMachine = new WebpMachine(); webpMachine.polyfillDocument()` everytime a route changes.

in the meantime, without workerizing webp-hero, i believe you can improve your performance for your app: i suggest you don't recreate a `new WebpMachine()` every time the route changes

instead, you should create only one webp machine during your app's initialization, and then you can call `polyfillDocument` whenever the images on the page might change

```js
// initially (only once) in your app
const webpMachine = new WebpMachine()

// run the polyfill whenever the route changes
async function onRouteChange() {
  await webpMachine.polyfillDocument()
}
```

this way, the webp-machine maintains a cache that operates between routes (so it doesn't decode the same image twice)

i suspect this optimization may improve your performance significantly, however, workerizing `webp.js` would greatly improve the perceived performance even moreso

**Comment by jnorthrup** [comment]
@developit

**Comment by chase-moskal** [comment]
@possoolutions, @jnorthrup

hey fellas, i think i might be misunderstanding this issue so far -- i'd like to ask some questions here for clarification

(edited 2019-11-09 adding answers from my sleuthing)

- [x] is this a performance/user-experience problem, or is webp-hero just plain broken in safari?

    it looks like emscripten/sdl, by default, is intercepting and disabling all normal keyboard input on the page

- [x] is this issue specific to safari? does it happen in other browsers when using the webp-hero demo in "?force" mode?

    no, it seems to reproduce in chrome too

- [x] does this issue involve server side rendering?
    - negative
    - webp-hero is not (yet?) designed to operate under nodejs

i do not have an apple device handy at the moment, so i would appreciate if anybody could contribute information about browser support in issue #8

**Comment by jnorthrup** [comment]
@chase-moskal we have react+nextjs
ok #8

**Comment by TingNgs** [comment]
I have the same issue in react + nextjs.
After i ran
`const webpMachine = new WebpMachine(); webpMachine.polyfillDocument()`
And input get disabled. even in chrome (force)

I found out from [webp-hero/libwebp/dist/webp.js](https://github.com/chase-moskal/webp-hero/blob/master/libwebp/dist/webp.js#L3)
added multi eventListener to the document and disabled keyboard event 
search  `if(!Module["doNotCaptureKeyboard"])`.

I simply solve this by adding 
`Module.doNotCaptureKeyboard = true;`
at line 20 and it works perfectly fine for me.

But i'm not sure it suit for different cases, hope it can help.

**Comment by chase-moskal** [comment]
taking another peek at this issue, i'm thinking we should probably add

`Module.doNotCaptureKeyboard = true`

to [webp-hero/libwebp/source/wrapping-end.js.txt](https://github.com/chase-moskal/webp-hero/blob/master/libwebp/source/wrapping-end.js.txt)

as @TingNgs indicates :+1: 

this will require a rebuild of libwebp from the dockerfile, and we ought to update the emscripten sdk and libwebp version while we're at it

**Comment by chase-moskal** [comment]
[a relevant github issue, emscripten 6476](https://github.com/emscripten-core/emscripten/issues/6476)

**Comment by chase-moskal** [comment]
it looks like issue #21 is a duplicate of this one

**Comment by chase-moskal** [comment]
[this stack answer](https://stackoverflow.com/a/55270572/1033426) says we can maybe set `SDL_SetHint(SDL_HINT_EMSCRIPTEN_KEYBOARD_ELEMENT, "#canvas");` instead of using `Module.doNotCaptureKeyboard`.. food for thought...

**Comment by Nican** [comment]
I spent way too much time debugging this issue. Here is my fix for anyone that just wants something to copy and paste and fix the problem:
```
import { WebpMachine, detectWebpSupport, defaultDetectWebpImage } from 'webp-hero';
import {Webp} from "webp-hero/libwebp/dist/webp.js"
export default class FixedWebpMachine extends WebpMachine {
    constructor() {
        const opts = {
            webp: new Webp(),
            webpSupport: detectWebpSupport(),
            detectWebpImage: defaultDetectWebpImage
        };
        opts.webp.Module.doNotCaptureKeyboard = true;
        super(opts);
    }
}
```
Thank you for maintaining this library, @chase-moskal.

**Comment by 06Games** [comment]
Any news on this subject? I also encounter this bug on IE11.
@Nican How am I supposed to put your fix in place? What files are required? Could you enlighten me a little on the subject...

**Comment by Nican** [comment]
@06Games 
Looks like this project had a major refactor, and the fix was directly integrated into here: 
https://github.com/signalstickers/signalstickers/blob/master/src/lib/convert-image.ts#L28

**Comment by 06Games** [comment]
OK, thanks

**Comment by chase-moskal** [comment]
today, i wanted to update the library and slip this fix in, however i can't get libwebp building anymore, see #31 -- so that new issue is currently blocking this one

the fix is in commit cda1bd687f34d99a8c5b65461f8ff2fa0be43525, but we can't release it until we can build it

**Comment by chase-moskal** [comment]
@06Games 

i've made progress on upgrading and refactoring the libwebp build, but we're now stuck on https://github.com/chase-moskal/webp-hero/issues/31, which itself is blocked by this funky emscripten wasm issue https://github.com/emscripten-core/emscripten/issues/10434

i'm waiting on response on the emscripten issue, once that's fixed, webp-hero will perform better, and we'll be able to merge this issue's fix
